# Cleaned_Bank_Reviews (BMO)

In [1]:
# Load packages
import numpy as np
import pandas as pd
import datatable as dt

## Load Tweets

In [2]:
# Load previously saved BMO tweets
df_bmo_tweets_one = pd.read_csv(r'@bmo_tweets.csv',encoding='utf-8') #keywords = "@bmo"
df_bmo_tweets_two = pd.read_csv(r'bmo_AND_bank_tweets.csv',encoding='utf-8') #keywords = "#bmo And bank" as "#bmo" itself will extract irrelevant tweets
print(df_bmo_tweets_one.shape,df_bmo_tweets_two.shape)

(59604, 10) (905, 10)


In [4]:
#combine two tweets dataframe
df_bmo_tweets = pd.concat([df_bmo_tweets_one, df_bmo_tweets_two], ignore_index=True)
print(type(df_bmo_tweets))
print(df_bmo_tweets.size)
print(df_bmo_tweets.head())

<class 'pandas.core.frame.DataFrame'>
605090
              user                       fullname             tweet-id  \
0  Tpyo9K7PUIEKlMC  อย่ามองตาได้ไหมถ้าเธอไม่หล่อ~  1098974331097378816   
1   ltopshbottom28           ˡᵃᵘʳ  🌼 BLOCK DE SDV  1098710034173870080   
2     Mareeeeeeech                            mr.  1098628200022847489   
3       AnimalGG15                       Animal15  1098390118317072384   
4       megami_901                             女神  1098628196533063680   

             timestamp                                          url  likes  \
0  2019-02-22 15:54:33  /Tpyo9K7PUIEKlMC/status/1098974331097378816      1   
1  2019-02-21 22:24:19   /ltopshbottom28/status/1098710034173870080      0   
2  2019-02-21 16:59:09     /Mareeeeeeech/status/1098628200022847489      1   
3  2019-02-21 01:13:05       /AnimalGG15/status/1098390118317072384      3   
4  2019-02-21 16:59:08       /megami_901/status/1098628196533063680    107   

   replies  retweets                     

# Why there's non English tweets???

## Load App Reviews

In [5]:
# Load previously saved BMO App Reviews
df_bmo_app = pd.read_csv(r'BMO_App_Reviews.csv', encoding ='utf-8')
print(df_bmo_app.size)

232139


In [6]:
#keep English reviews only
df_bmo_app= df_bmo_app[df_bmo_app['Language'] == 'English']
print(df_bmo_app.size)

213118


### select/change column names from app review

In [7]:
df_bmo_app=df_bmo_app[['Date', 'Author', 'App Name', 'App Store', 'Rating', 'Body']]
df_bmo_app.head()

Date               Author            App Name    App Store  Rating  \
0  2019-03-21     Sparrow Crandall  BMO Mobile Banking  Google Play       1   
1  2019-03-21      Coconut Licious  BMO Mobile Banking  Google Play       5   
2  2019-03-21       Jeremiah David  BMO Mobile Banking  Google Play       5   
4  2019-03-20  Etienne Matuszewski  BMO Mobile Banking  Google Play       1   
5  2019-03-20        Stefdax Bijou  BMO Mobile Banking  Google Play       1   

                                                Body  
0  Honestly this app makes my blood boil to the p...  
1      Super user friendly and has everything I need  
2  above average excellent appt does everything I...  
4  According to Protectstar anti-spying apps, the...  
5  Always need to have Wi-Fi to get my bank info,...

In [8]:
#rename cols
df_bmo_app=df_bmo_app.rename(index=str, columns={'App Store': 'Platform', 'Body': 'Review', 'Author': 'User', 'Rating': 'Rating/Likes'})

In [9]:
#converting Date from 'object' to datetime
df_bmo_app['Date'] = pd.to_datetime(df_bmo_app['Date'])
print(df_bmo_app.size)

55596


In [10]:
#identify sourses
df_bmo_tweets['Platform'] = 'Twitter'
df_bmo_tweets['App Name'] = 'BMO'
# Check the top records
print(df_bmo_tweets.head())

              user                       fullname             tweet-id  \
0  Tpyo9K7PUIEKlMC  อย่ามองตาได้ไหมถ้าเธอไม่หล่อ~  1098974331097378816   
1   ltopshbottom28           ˡᵃᵘʳ  🌼 BLOCK DE SDV  1098710034173870080   
2     Mareeeeeeech                            mr.  1098628200022847489   
3       AnimalGG15                       Animal15  1098390118317072384   
4       megami_901                             女神  1098628196533063680   

             timestamp                                          url  likes  \
0  2019-02-22 15:54:33  /Tpyo9K7PUIEKlMC/status/1098974331097378816      1   
1  2019-02-21 22:24:19   /ltopshbottom28/status/1098710034173870080      0   
2  2019-02-21 16:59:09     /Mareeeeeeech/status/1098628200022847489      1   
3  2019-02-21 01:13:05       /AnimalGG15/status/1098390118317072384      3   
4  2019-02-21 16:59:08       /megami_901/status/1098628196533063680    107   

   replies  retweets                                               text  \
0        0 

### In the df_tweets, we are going to keep timestamp, user, likes, text Add App Name and Platform.Then, we will remove duplicates.

In [11]:
#rename cols
df_bmo_tweets=df_bmo_tweets.rename(index=str, columns={'timestamp':'Date','user':'User','likes':'Rating/Likes','text':'Review'})

In [12]:
#change col order
df_bmo_tweets=df_bmo_tweets[['Date','User','App Name','Platform','Rating/Likes','Review']]
df_bmo_tweets.head()

Date             User App Name Platform  Rating/Likes  \
0  2019-02-22 15:54:33  Tpyo9K7PUIEKlMC      BMO  Twitter             1   
1  2019-02-21 22:24:19   ltopshbottom28      BMO  Twitter             0   
2  2019-02-21 16:59:09     Mareeeeeeech      BMO  Twitter             1   
3  2019-02-21 01:13:05       AnimalGG15      BMO  Twitter             3   
4  2019-02-21 16:59:08       megami_901      BMO  Twitter           107   

                                              Review  
0  ขอยาสเกลียสแท้กนิดนุงนะ5555 #ประยุทธ์ไม่กล้าดีเบต  
1                                  Follow me pleasee  
2                                                NaN  
3  Gente!! Me ayudan con 15rt para que @z_slopy m...  
4  鈴(りん)でちゅ\n今からジャンプちゅるでちゆ\n見ててくだちゃいね。pic.twitter...

In [13]:
#converting Date from 'object' to datetime
df_bmo_tweets['Date'] = pd.to_datetime(df_bmo_tweets['Date'])

### convert to lowercase, keep tweets contains 'bmo'

In [14]:
# Convert them to lowercase
df_bmo_tweets['Review'] = df_bmo_tweets['Review'].str.lower()
print(df_bmo_tweets.size)

363054


In [15]:
# Remove duplicates
df_bmo_tweets= df_bmo_tweets.drop_duplicates(subset=['Review'])
print(df_bmo_tweets.size)

351906


In [16]:
# check NaN
df_bmo_tweets.isna().sum()

Date            0
User            0
App Name        0
Platform        0
Rating/Likes    0
Review          1
dtype: int64

In [17]:
# Keep tweets that contain 'bmo'
df_bmo_tweets = df_bmo_tweets[df_bmo_tweets['Review'].str.contains('bmo',na=False)]
print(df_bmo_tweets.size)

183048


### dropping all users that have 'BMO' in their username as we are interested only in clients and not bank's responses.

In [18]:
df_bmo_tweets = df_bmo_tweets[~df_bmo_tweets.User.str.contains('BMO', na = False)]
print(df_bmo_tweets.size)

166398


### Now merge the two datasets.

In [19]:
# Appending two dataframes
df_combined = pd.concat([df_bmo_tweets, df_bmo_app], ignore_index = True)
print(df_combined.head())
print('\n--Total Number of De-duplicated Tweets that contain \'bmo\': {} --\n'.format(len(df_bmo_tweets)))
print('\n--Total Number of English App Reivews: {} --\n'.format(len(df_bmo_app)))
print('\n--Total Number of Comments: {} --'.format(len(df_combined)))

                 Date           User App Name Platform  Rating/Likes  \
0 2019-01-16 23:33:33  WisconsinHerd      BMO  Twitter             1   
1 2019-01-16 22:51:53  Ryan_King_EDU      BMO  Twitter            40   
2 2019-01-16 22:13:13        DrCeitm      BMO  Twitter             0   
3 2019-01-16 21:28:24   Catalyst_yyc      BMO  Twitter             1   
4 2019-01-16 20:55:09      IE_Canada      BMO  Twitter             0   

                                              Review  
0   congrats to you and your school!\n@bmoharrisbank  
1  i knew my puzzle skills and jumper would event...  
2  once @bmo showed me they are win. they were so...  
3  getting close to our agriculture conference. w...  
4  .@bmo creates bmo private wealth canada and as...  

--Total Number of De-duplicated Tweets that contain 'bmo': 27733 --


--Total Number of English App Reivews: 9266 --


--Total Number of Comments: 36999 --


### After consolidating the dataset we end up with 36999 reviews to work with for BMO

In [20]:
# dataframe to csv file
df_combined.to_csv(r'bmo_consolidated.csv', index=False)

In [29]:
### check English
from langdetect import detect

In [ ]:
lang = []    
for r in df_combined['Review']:
    lang = detect(r)
    df_combined['Languagereveiw'] = lang